### PR TITLE
[feat] Move from `setup.py` and `setup.cfg` to `pyproject.toml`

### PR DESCRIPTION
**Pull request by fpgmaas**
Since [PEP621](https://peps.python.org/pep-0621/), `pyproject.toml` is the standard of specifying metadata for Python projects. This PR proposes to move the project to that new standard. See also [here](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html).

You can build the wheel with:

```sh
pip install build wheel
python -m build --wheel
```

Please let me know your thoughts on the proposal. Happy to make changes where required, and if you are not interested in the move to `pyproject.toml`, feel free to close.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -35,12 +35,12 @@ jobs:
         id: restore-cache
         with:
           path: ${{ env.pythonLocation }}
-          key: python-dependencies-${{ matrix.os }}-${{ matrix.python-version }}-${{ hashFiles('setup.py') }}-${{ hashFiles('requirements.txt') }}-${{ env.pythonLocation }}
```
Comment (tomaarsen) [inline .github/workflows/tests.yml:38]: Very thorough, well done.

**Review by tomaarsen — COMMENTED** [review]
Personally, I'm quite a fan of `pyproject.toml`, so I can imagine myself moving forward with this.
I'll have to verify that everything still works as expected, primarily regarding building wheels etc.

Do you know whether the [MANIFEST.in](https://github.com/UKPLab/sentence-transformers/blob/master/MANIFEST.in) is still taken into consideration when building via `pyproject.toml`? Alternatively, can we package that information in `pyproject.toml` instead? I always thought it was a bit of a waste to have to add a completely new file just to include one markdown file in the wheel.

**Comment by fpgmaas** [comment]
> Personally, I'm quite a fan of `pyproject.toml`, so I can imagine myself moving forward with this. I'll have to verify that everything still works as expected, primarily regarding building wheels etc.
> 
> Do you know whether the [MANIFEST.in](https://github.com/UKPLab/sentence-transformers/blob/master/MANIFEST.in) is still taken into consideration when building via `pyproject.toml`? Alternatively, can we package that information in `pyproject.toml` instead? I always thought it was a bit of a waste to have to add a completely new file just to include one markdown file in the wheel.

Good question, yes I believe the `MANIFEST.in` is included;

```
➜ python -m build --wheel
...
Successfully built sentence_transformers-3.1.0.dev0-py3-none-any.whl

➜ zipinfo dist/sentence_transformers-3.1.0.dev0-py3-none-any.whl
Archive:  dist/sentence_transformers-3.1.0.dev0-py3-none-any.whl
Zip file size: 229841 bytes, number of entries: 101
...
-rw-r--r--  2.0 unx     9638 b- defN 24-Jun-27 06:11 sentence_transformers/model_card_template.md
...
```

I believe at the time of writing there is no way to include this info in `pyproject.toml` with setuptools, see also [here](https://github.com/pypa/setuptools/issues/3341).

It would be possible if we use for example PDM as a build backend, see [here](https://backend.pdm-project.org/build_config/). I personally like using a tool like that, since they also provide some other utilities like lock files, easily creating of virtual environment, etc.

Let me know if you want me to implement an alternative PR that uses PDM to compare solutions.

**Comment by fpgmaas** [comment]
I see checks are failing, sorry! I will work on this a bit more in my fork and update the PR soon.

**Comment by tomaarsen** [comment]
All good; I think we're best off to stick with the basics/defaults as they're more familiar and less likely to break in the future.

It looks like installing `sentence-transformers` in the "external dependencies" section of the CI is causing issues.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -30,24 +30,19 @@ jobs:
...
+            ~/.cache/pip
+          key: python-dependencies-${{ matrix.os }}-${{ matrix.python-version }}-${{ hashFiles('pyproject.toml') }}
 
-      - name: Install external dependencies on cache miss
+      - name: Install dependencies
         run: |
-          python -m pip install --no-cache-dir --upgrade pip
-          python -m pip install --no-cache-dir -r requirements.txt
-          python -m pip install --no-cache-dir pytest
-        if: steps.restore-cache.outputs.cache-hit != 'true'
```
[the remark this answers is not in the document]
Comment (fpgmaas) [reply .github/workflows/tests.yml:45]: Good catch; I think I solved that now. See e.g. [this run](https://github.com/fpgmaas/sentence-transformers/actions/runs/9796746920/job/27052472245)

**Pull request merged**